### ISSUE TITLE
Handling of bad data from sources

### ISSUE THREAD

**Issue by elliottwilliams** [propershark/shark#9]
This route object broke my decoder:

```
[12] = Object {
  Object = 7 key/value pairs {
    [0] = {
      key = "short_name"
      value = String (String = "19")
    }
    [1] = {
      key = "color"
      value = Null
    }
    [2] = {
      key = "description"
      value = String (String = "Inner Loop")
    }
    [3] = {
      key = "code"
      value = Null
    }
    [4] = {
      key = "stations"
      value = Null
    }
    [5] = {
      key = "path"
      value = Null
    }
    [6] = {
      key = "name"
      value = String (String = "Inner Loop")
    }
  }
}
```

I remember you noticing this null-ish route before, but I'm wondering what you plan to do with things like this. Will you enforce that data coming in conforms to object property types before sending them out as events? Or do I need to filter the Shark input stream for things like this in the client?

**Comment by faultyserver** [comment]
I think client side is the right place to do this. I can see filtering on the server-side being too coarse, potentially masking data that _could_ be relevant to the users (new routes that aren't yet active, etc). It's minuscule and not really an issue with CityBus; but again, generality is good.

Also, ensuring that masked data is fully-masked (all references to it are nulled as well) could lead to more complexity and potential issues than dealing with it client side.

#### Example

Given that a route like this exists, and it gets filtered on the server side: it needs to be masked from a) the activate cache and b) publishing updates. Say the route loses information while it is already active: then a `deactivate` event has to be pushed, even though it could still be active (say DoubleMap `/routes` goes down for 20 minutes like it did on Tuesday).

Now, let's say a vehicle was traveling on that route when it lost information. If we mask the newly-nulled-and-deactivated route, then this vehicle is no longer "valid" in that it references a route that essentially doesn't exist.

While you could use `meta.last_event` to pull the old information about the route, the fact that it's not in the `agency.routes` topic (since it was deactivated) means that there is no direct path the user can take from a new session to view that route, even though a vehicle that they may be interested in is still traveling on it.

On the other hand, let's assume it gets filtered on the client side: you can grey it out/have some alert bar that says "this route is currently under maintenance and is not providing all of its information" or something that indicates there is an issue with the sourcing of information. You can also still use the information that _is_ provided to fill in references elsewhere (should they exist).

In the case of a vehicle traveling on the route, it's references are preserved, and users can take a direct path to the route from anywhere to see the route's information.

The implementation difficulty I can see on the client side would be trying to avoid nulling out information you already know (e.g., path; which, in fairness, might have been nulled out because it is no longer valid).

#### Another (Simpler) Example

Say an agency doesn't provide the "color" for one of it's routes. We could easily say "color isn't an important attribute", but that's not generic. Being generic would mean trying to determine if that route should get filtered based on configuration options for each type of object (at a minimum) or, worse, each type of event. <sub>I make this sound like something terrible when it really isn't, but my point is that the writing the configuration files would be rather tedious and nit-picky.</sub>

On the client side, however, you could receive a route with no color and just use some default color  in the display instead. Where this color comes from doesn't matter, but is certainly a UI detail, not a data-level detail.

**tl;dr:** Filtering on the client side seems better because you have more control of what gets filtered , why it gets filtered, and how you respond. Additionally, dealing with null values on the client is useful for agencies that don't provide a given attribute for one object out of a list.

I'm open to discussion, but I don't see any benefit in handling this on the server side.

**Comment by elliottwilliams** [comment]
That's compelling. I'll sleep on it and probably close this tomorrow..
